src/http/gateway/resolver.js

```diff
@@ -0,0 +1,114 @@
+'use strict'
+
+const mh = require('multihashes')
+const promisify = require('promisify-es6')
+const eachOfSeries = require('async/eachOfSeries')
+const debug = require('debug')
+const log = debug('jsipfs:http-gateway:resolver')
+log.error = debug('jsipfs:http-gateway:resolver:error')
+
+const html = require('./utils/html')
+const PathUtil = require('./utils/path')
+
+const INDEX_HTML_FILES = [ 'index.html', 'index.htm', 'index.shtml' ]
+
+const resolveDirectory = promisify((ipfs, path, multihash, callback) => {
+  if (!callback) {
+    callback = noop
+  }
+
+  mh.validate(mh.fromB58String(multihash))
+
+  ipfs
+   .object
+   .get(multihash, { enc: 'base58' })
+   .then((DAGNode) => {
+     const links = DAGNode.links
+     const indexFiles = links.filter((link) => INDEX_HTML_FILES.indexOf(link.name) !== -1)
+
+     // found index file in links
+     if (indexFiles.length > 0) {
+       return callback(null, indexFiles)
+     }
+
+     return callback(null, html.build(path, links))
+   })
```

Comment:
Mixing Promises and Callbacks has been historically a recipe for problems. Here the error is not being caught (promise) and the promise is possibly returning before the callback fires.

Let's continue using Callbacks everywhere when possible.